### PR TITLE
Modify unique to return sizes when dim is zero-length

### LINKED ISSUE #75730: `torch.unique` for zero-length `dim` doesn't preserve the input size
### Description

Usually, when applying `torch.unique` to a specific dimension, `torch.unique(dim=x)` returns output that, except for dim `x`, matches the size of the input tensor.
However, in the case where dim `x` in the input tensor has length zero, the input dimensions are discarded, and `torch.unique` output erroneously has size `[0]`. This happens regardless of the shape of the input tensor.

### Steps to reproduce:

```python
import torch

# Create a multi-dimensional tensor, with a single zero-length dimension.
input_tensor = torch.randint(3, [0, 4])

# Call `torch.unique` on the tensor, along the zero-length dimension.
input_tensor.unique(dim=0)

# Output:
# torch.tensor([], dtype=torch.int64)
```

### Expected behaviour

Except for the dimension supplied to `dim`, the dimensions of the output of `unique` should always match the input.

```python
torch.randint(3, [0, 4]).unique(dim=0)

# Output:
# torch.tensor([], size=(0, 4), dtype=torch.int64)




# Or, for a differently-sized input tensor:
torch.randint(3, [2, 0, 9]).unique(dim=1)

# Output:
# torch.tensor([], size=(2, 0, 9), dtype=torch.int64)
```

### Actual behaviour

Calling `torch.unique(dim=x)`, where `x` is a zero-length dimension, returns a tensor of size `[0]`, regardless of any other dimensions of the input tensor.

```python
torch.randint(3, [0, 4]).unique(dim=0)

# Output:
# torch.tensor([], dtype=torch.int64)




torch.randint(3, [2, 0, 9]).unique(dim=1)

# Output:
# torch.tensor([], dtype=torch.int64)
```

### Why do I expect the behaviour in "Expected behaviour"?

When acting along a zero-length dimension, other Pytorch methods preserve the dimensions of the input tensor, rather than returning a tensor of size `[0]`.

Such as `sort`:

```python
import torch

# Create a multi-dimensional tensor, with a single zero-length dimension.
input_tensor = torch.randint(3, [0, 4])

# Call `torch.sort` on the tensor, along the zero-length dimension.
input_tensor.sort(dim=0).values

# Output:
# torch.tensor([], size=(0, 4), dtype=torch.int64)
```

And `cat`:

```python
import torch

# Create a multi-dimensional tensor, with a single zero-length dimension.
input_tensor = torch.randint(3, [0, 4])

# Call `torch.sort` on the tensor, along the zero-length dimension.
torch.cat([input_tensor]*2, dim=0)

# Output:
# torch.tensor([], size=(0, 4), dtype=torch.int64)
```

### Proposed changes

1. Replace `{0}` on [this line](https://github.com/pytorch/pytorch/blob/89486821edf0aad15da7003c8f10e984b53f762e/aten/src/ATen/native/Unique.cpp#L185) with `sizes`

2. Replace `{0}` on [this line](https://github.com/pytorch/pytorch/blob/89486821edf0aad15da7003c8f10e984b53f762e/aten/src/ATen/native/cuda/Unique.cu#L120) with `sizes`

### Versions

Collecting environment information...
PyTorch version: 1.12.0a0
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.23.0
Libc version: glibc-2.27

Python version: 3.9.0 (default, Apr 10 2022, 10:48:11)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-72-generic-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: NVIDIA GeForce 940MX
Nvidia driver version: 470.103.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] torch==1.12.0a0+gitunknown
[conda] Could not collect

### PR DESCRIPTION
**Pull request by ryandaryl**
Fixes #75730

Previously, `torch.unique` returned a tensor of size `[0]` when the dimension supplied to `dim` is zero length. Instead, I've changed `torch.unique` to return a tensor matching the input size.

Changes:
- Modify `torch.unique` to return an an empty tensor of size `sizes` as the values output, when `dim` is zero length in the input tensor.
- Make the same change in the CUDA implementation of `torch.unique`.
- Update `test_unique_dim` to expect an empty tensor with size matching the input tensor.

### CODE REVIEW

**Review by mruberry — APPROVED** [review]
Thanks @ryandaryl!

As a note for the project, in the future we should update unique's OpInfo to include a NumPy reference operation and programmatically validate its behavior is consistent with NumPy.
